packages/ckeditor5-ui/src/toolbar/block/blocktoolbar.ts

```diff
@@ -497,3 +626,39 @@ export default class BlockToolbar extends Plugin {
 		return toPx( editableRect.width + offset );
 	}
 }
+
+/**
+ * Returns the closest scrollable parent element of the given element.
+ *
+ * @param element - The element for which to find the closest scrollable parent.
+ * @returns The closest scrollable parent element, or `null` if no scrollable parent is found.
+ */
+function closestScrollableParent( element: HTMLElement ): HTMLElement | null {
+	if ( isScrollable( element ) ) {
+		return element;
+	}
+
+	let parent = element.parentElement;
+	let depth = 0;
+
+	// Check up to 6 levels of parent elements to prevent performance issues.
+	while ( parent && !isScrollable( parent ) ) {
+		if ( depth++ >= 6 ) {
```

Comment:
Why such an arbitrary number? This could be hard to debug if it happens. If the performance is an issue we could store it and not search for it every time. Did you notice some performance issues without this limitation or is it an early optimization without real data that it is required?